File: Python/4949 - Balanced world.py
def check(string):
    stack = []
    for elem in string:
        if elem == '(':
            stack.append(elem)

        elif elem == '[':
            stack.append(elem)

        elif elem == ')':
            if not stack or stack[-1] != '(':
                return "no"
            else:
                stack.pop()

        elif elem == ']':
            if not stack or stack[-1] != '[':
                return "no"
            else:
                stack.pop()

    if stack:
        return "no"
    else:
        return "yes"


while True:
    s = input()
    if s == '.':
        break
    print(check(s))

File: C++/4949 - Balanced World.cpp
#include<bits/stdc++.h>

using namespace std;


string check(string line) {
    stack<char> stk;
    
    for (auto elem : line) { // iterate all characters in line
        
        if (elem == '(') stk.push(elem);
    
        else if (elem == '[') stk.push(elem);
        
        else if (elem == ')') {
            // when stack is empty or unable to match
            if (stk.empty() || stk.top() != '(') return "no";
            // proceed to the next char
            else stk.pop();
        }
        else if (elem == ']') {
            // when stack is empty or unable to match
            if (stk.empty() || stk.top() != '[') return "no";
            // proceed to the next char
            else stk.pop();
        }
    }

    // if there is remaining character(s)
    if (!stk.empty()) return "no";
    
    // stack is empty = all parenthesises are matched
    else return "yes";
    
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(nullptr);

    string line = "";

    while (line != ".") {
        getline(cin, line);
        cout << check(line) << "\n";
    }

    return 0;
}
